### DISCUSSION TITLE
NextJS Page co-location in non-frame specific apps.

### DISCUSSION THREAD

**Discussion by dalechyn** [wevm/frog-archived#20] [Q&A] [answered]
It is known that a "Frame" itself is a combination of `meta` tags that are put into the page's `<head>` element, and the need to separate the page and a "Frame" is kinda breaking the semantic idea behind such being meta tags.

In other words, If I cast a link to my project that would look like this: `example.com`, I want to see a frame derived from that specific link, and not have `example.com/api` with a `browserLocation` parameter.
I want users to easily share a link of any page on the website, and if such has frame meta tags set, it should be rendered in the same path.

As of now, `frog` forces me to build frames under a specific location which should differ from the page location as by [convention](https://nextjs.org/docs/app/building-your-application/routing/route-handlers#convention), I cannot place `route` and `page` in the same layout segment.

It makes augmented sharing impossible. Instead of having a user to share `example.com/post/frog-released`, they would have to cast `example.com/api/post/frog-released`.

### Considerations
- `frog/next` package with a helper `metadata` that would take the `Frog` instance and so some magic to have it used in the `export const metadata = {...}`
- To build a parallel page route where a server component would use `frog/next`'s `FrameMirror` component that would take `url: string` as a parameter for the route handler path to mirror. Post requests would of course be handled in a separate segment level.
  Pseudo-code example:
  ```tsx
  // un-changed example from docs, but with a different basePath
  // app/api/frames/[[...routes]]/route.tsx
  /** @jsxImportSource frog/jsx */
  import { Button, Frog, TextInput } from 'frog'
  import { handle } from 'frog/vercel'
   
  const app = new Frog({ 
    basePath: '/api/frames',
    // Supply a Hub API URL to enable frame verification.
    // hubApiUrl: 'https://api.hub.wevm.dev',
  })
   
  // Uncomment to use Edge Runtime
  // export const runtime = 'edge'
   
  app.frame('/', (c) => {
    const { buttonValue, status } = c
    return c.res({
      image: (
        <div style={{ color: 'white', display: 'flex', fontSize: 60 }}>
          {status === 'initial' ? (
            'Select your fruit!'
          ) : (
            `Selected: ${buttonValue}`
          )}
        </div>
      ),
      intents: [
        <Button value="apple">Apple</Button>,
        <Button value="banana">Banana</Button>,
        <Button value="mango">Mango</Button>
      ]
    })
  })
   
  export const GET = handle(app)
  export const POST = handle(app)
  ```
  ```tsx
  // app/@frog/page.tsx
  import { FrameMirror } from 'frog/next'
  
  // `FrameMirror` would make a `GET` request to the passed `url`,
  // extract the frame's `meta` tags and render them in here,
  // with the `postUrl` to be `/api/frames`
  export default async () => <FrameMirror url="/api/frames" />
  ```
- `exportFrameMetadata` util that would accept `url: string` as parameter and would simply extract the frame meta tags to be put in `export const metadata`.

**Comment by dalechyn** [comment]
Here's a workaround:

1. Place a function `extractFrameMetaTags` somewhere in your repo with the next contents:
  ```tsx
  export async function extractFrameMetaTags(
    url: string,
  ): Promise<Record<string, string>> {
    try {
      const res = await fetch(url)
      const text = await res.text()
  
      const match = text.matchAll(/<meta property="fc:(\S+)" content="(\S+)"/gm)
      if (!match) return {}
  
      const metaTags: Record<string, string> = {}
      for (const matchGroup of match) {
        const key = matchGroup[1]
        const value = matchGroup[2]
        metaTags[`fc:${key}`] = value
      }
  
      return metaTags
    } catch (error) {
      throw new Error(`Failed to extract frame meta tags from ${url}: ${error}`)
    }
  }
  ```
  This method will make a `GET` request to the endpoint you pass and will try to extract the `fc:*` properties in meta if such are present.
2. In your `page` file, use `generateMetadata` function to extract such properties from the frog!
   ```tsx
  // `APP_URL` is a constant that represents the url to your app.
  // in this example, i have modified `basePath` for frog to `/api/frog`.
  export async function generateMetadata(): Promise<Metadata> {
    const frameTags = await extractFrameMetaTags(`${APP_URL}/api/frog`)
      return {
        other: frameTags,
      }
  }
  ```
I've tested this in `frame.js` debugger and it works!
![Screenshot 2024-02-29 at 16 25 30](https://github.com/wevm/frog/assets/35642018/48a832f7-3121-4af1-b2a9-8b7aa03c8fb9)

**Comment by tmm** [comment, marked as answer]
Just commenting that Frog now has a first-class support for this, added in #31

https://frog.fm/platforms/next#bonus-page--frame-co-location
